### ISSUE TITLE
Indicate which files / directories have been added, modified, or deleted

### ISSUE THREAD

**Issue by kumar303** [mozilla/addons-code-manager#413]
User story:

> As a reviewer, I want to quickly see which files have been added, modified, or deleted in a new version so that I only need to review the relevant files

As an example, the current file viewer highlights the modified files / directories in yellow but we don't need to do it exactly this way. 

Here, the `install.rdf` file was modified and the `META-INF` directory contains modified files:

<img width="1192" alt="Screenshot 2019-03-18 12 19 24" src="https://user-images.githubusercontent.com/55398/54549459-15abe200-4978-11e9-9df1-4ab1a85e81c8.png">

Here, the `META-INF` directory and the `manifest.json` file were deleted:

<img width="1193" alt="Screenshot 2019-03-18 12 21 35" src="https://user-images.githubusercontent.com/55398/54549624-73d8c500-4978-11e9-92a1-1d3a95c7e2c8.png">

https://reviewers.addons-dev.allizom.org/en-US/firefox/files/compare/262229...262228/#top

**Comment by willdurand** [comment]
Has the API landed for that?

**Comment by kumar303** [comment]
Is this blocked? I thought we could use `mode` in the `diff` response of [the compare endpoint](https://addons-server.readthedocs.io/en/latest/topics/api/reviewers.html#compare) to figure out if the file was added / deleted.

**Comment by willdurand** [comment]
We need confirmation from @EnTeQuAk but I remember a discussion about the `file.entries` containing the information about the added/deleted/renamed files.

**Comment by EnTeQuAk** [comment]
> Is this blocked? I thought we could use mode in the diff response of the compare endpoint to figure out if the file was added / deleted.

That is actually the API. The only part that hasn't landed yet is support for detecting copy and renames but that's a nice-to-have and doesn't exist in the current tool.

If you need that information in `files.entries` too, let me know but that feels somehow like duplication to me.

**Comment by willdurand** [comment]
I am confused because I thought the Compare API returned a `diff` for a single file, but it looks like we want the `mode` for all the files here, hence the discussion I still think we had a while ago about having the Compare API returning `file.entries` with all the files for the two compared versions + the `mode` for each file (mode being: added/deleted/renamed).

**Comment by EnTeQuAk** [comment]
Some background: The diff api can very happily return a diff for all files (needs a change but we explicitly strict the diff currently, the diff is always generated for all files in the background).

That being said, if that restriction is lifted the response will probably become very huge. Which may not be a network/performance related problem because of gzipping but it'll most certainly consume huge amounts of client-side memory.

So, I do think that @willdurand is right that the mode should probably be exposed in the `entries` list. I'll explore how much work that'll be.

**Comment by willdurand** [comment]
Old IRC log: https://mozilla.logbot.info/amo/20190222#c15999496-c15999501

**Comment by EnTeQuAk** [comment]
PR is up: https://github.com/mozilla/addons-server/pull/11124

**Comment by kumar303** [comment]
@EnTeQuAk ok, I can confirm that the `diff` output won't work for this feature. It's currently missing the added and deleted part of the diff. Can you confirm that what you're adding in https://github.com/mozilla/addons-server/pull/11124 will give us access to added and deleted files?

Here is a test I did in dev: https://code.addons-dev.allizom.org/en-US/compare/616645/versions/1655530...1655531/

I was expecting the `diff` below to include the added file (`scripts/content.js`) and the deleted file (`scripts/background.js`) but it contains neither.

Here is the API response from `https://addons-dev.allizom.org/api/v4dev/reviewers/addon/616645/versions/1655530/compare_to/1655531/`:

<details>
<summary>JSON</summary>

```json
{
  "id": 1655531,
  "channel": "listed",
  "compatibility": {
    "firefox": {
      "min": "48.0",
      "max": "*"
    }
  },
  "edit_url": "https://addons-dev.allizom.org/en-US/developers/addon/git-add-delete-example/versions/1655531",
  "is_strict_compatibility_enabled": false,
  "license": {
    "id": 3338,
    "is_custom": false,
    "name": {
      "en-US": "Mozilla Public License, version 2.0"
    },
    "text": {
      "en-US": "This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at <a rel=\"nofollow\" href=\"https://outgoing.stage.mozaws.net/v1/82bb400dfa1ad16e455413fa870b7798a7796ec70a5c8a36cbf24b7c08635a49/http%3A//mozilla.org/MPL/2.0/\">http://mozilla.org/MPL/2.0/</a>."
    },
    "url": "http://www.mozilla.org/MPL/2.0/"
  },
  "release_notes": null,
  "reviewed": null,
  "version": "2.0.0",
  "file": {
    "id": 376192,
    "created": "2019-04-03T20:47:26Z",
    "hash": "sha256:d7d845bb27c13a2a9b9b80b11c922e0fe08385d996e513cf6dcef2d0305c2832",
    "is_restart_required": false,
    "is_webextension": true,
    "is_mozilla_signed_extension": false,
    "platform": "all",
    "size": 508,
    "status": "unreviewed",
    "url": "https://addons-dev.allizom.org/firefox/downloads/file/376192/git_adddelete_example-2.0.0-fx.xpi?src=",
    "permissions": [
      "*://*.mozilla.org/*"
    ],
    "diff": [
      {
        "path": "manifest.json",
        "size": 235,
        "lines_added": 11,
        "lines_deleted": 4,
        "is_binary": false,
        "mode": "M",
        "hunks": [
          {
            "header": "@@ -1,8 +1,15 @@",
            "old_start": 1,
            "new_start": 1,
            "old_lines": 8,
            "new_lines": 15,
            "changes": [
              {
                "content": "{",
                "type": "normal",
                "old_line_number": 1,
                "new_line_number": 1
              },
              {
                "content": "  \"manifest_version\": 2,",
                "type": "normal",
                "old_line_number": 2,
                "new_line_number": 2
              },
              {
                "content": "  \"name\": \"Git Add/Delete Example\",",
                "type": "normal",
                "old_line_number": 3,
                "new_line_number": 3
              },
              {
                "content": "  \"version\": \"1.0.0\",",
                "type": "delete",
                "old_line_number": 4,
                "new_line_number": -1
              },
              {
                "content": "  \"background\": {",
                "type": "delete",
                "old_line_number": 5,
                "new_line_number": -1
              },
              {
                "content": "    \"scripts\": [\"scripts/background.js\"]",
                "type": "delete",
                "old_line_number": 6,
                "new_line_number": -1
              },
              {
                "content": "  }",
                "type": "delete",
                "old_line_number": 7,
                "new_line_number": -1
              },
              {
                "content": "  \"version\": \"2.0.0\",",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 4
              },
              {
                "content": "  \"content_scripts\": [",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 5
              },
              {
                "content": "    {",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 6
              },
              {
                "content": "      \"matches\": [",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 7
              },
              {
                "content": "        \"*://*.mozilla.org/*\"",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 8
              },
              {
                "content": "      ],",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 9
              },
              {
                "content": "      \"js\": [",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 10
              },
              {
                "content": "        \"scripts/content.js\"",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 11
              },
              {
                "content": "      ]",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 12
              },
              {
                "content": "    }",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 13
              },
              {
                "content": "  ]",
                "type": "insert",
                "old_line_number": -1,
                "new_line_number": 14
              },
              {
                "content": "}",
                "type": "normal",
                "old_line_number": 8,
                "new_line_number": 15
              }
            ]
          }
        ],
        "old_path": "manifest.json",
        "parent": "4bca7812731f507b90f8b057381064278a0bb5ab",
        "hash": "0bf8949f5d6ad90fd2bb139598fe3a34e984613f",
        "new_ending_new_line": true,
        "old_ending_new_line": true
      }
    ],
    "entries": {
      "manifest.json": {
        "depth": 0,
        "filename": "manifest.json",
        "sha256": "7211c209e582413603f0df10fbcd5133f3dff1875cd7107682a890d59b740ce6",
        "mime_category": "text",
        "mimetype": "application/json",
        "path": "manifest.json",
        "size": 235,
        "modified": "2019-04-03T20:47:27Z"
      },
      "scripts": {
        "depth": 0,
        "filename": "scripts",
        "sha256": "",
        "mime_category": "directory",
        "mimetype": "application/octet-stream",
        "path": "scripts",
        "size": null,
        "modified": "2019-04-03T20:47:27Z"
      },
      "scripts/content.js": {
        "depth": 1,
        "filename": "content.js",
        "sha256": "62adc253789bf336abd3b8b34df682565182cbd2bb4c82625bdee77630e77e9d",
        "mime_category": "text",
        "mimetype": "application/javascript",
        "path": "scripts/content.js",
        "size": 31,
        "modified": "2019-04-03T20:47:27Z"
      }
    },
    "selected_file": "manifest.json",
    "download_url": "https://addons-dev.allizom.org/en-US/reviewers/download-git-file/1655531/manifest.json/"
  },
  "validation_url": "/en-US/developers/addon/git-add-delete-example/file/376192/validation",
  "validation_url_json": "/en-US/developers/addon/git-add-delete-example/file/376192/validation.json",
  "has_been_validated": true,
  "addon": {
    "id": 616645,
    "slug": "git-add-delete-example",
    "name": {
      "en-US": "Git Add/Delete Example"
    },
    "icon_url": "https://addons-dev-cdn.allizom.org/static/img/addon-icons/default-64.png"
  }
}
```

</details>

**Comment by kumar303** [comment]
@EnTeQuAk just as a data point for your own queue, this is the only high priority issue we're currently blocked on.

**Comment by kumar303** [comment]
The API changes have landed: https://github.com/mozilla/addons-server/pull/11124

**Comment by bobsilverberg** [comment]
I have noticed something with the data, which is probably not a bug, but something we'll probably have to deal with.

It seems that a file can be changed without having any visible changes made inside the file, i.e., no lines added or deleted. I guess this can happen if the modification date for the file is updated without any actual changes made inside the file. I believe that we would want to treat that file as *unchanged*, so we'll need to deal with that when loading the data into our internal `Version`.

Here are a couple of examples:
https://code.addons-dev.allizom.org/en-US/compare/502955/versions/1541786...1541798/?path=META-INF%2Fmozilla.rsa
- that is a file that is marked as changed in `entries` (`status == 'M'`), but shows no changes in the UI (0 lines added, 0 lines deleted).

https://code.addons-dev.allizom.org/en-US/compare/502955/versions/1541786...1541798/?path=background.js
- that is a file that has no changes, and is marked with a `status` of `''`. In the UI this appears with the message "No differences".

In terms of displaying an indicator in the file tree of whether the file was modified, and also in terms of whether this file should be jumped to when navigating between changes via a keyboard shortcut (see #615), I believe that we should treat the first case as a file without any changes, but I could be convinced otherwise.

cc @kumar303 @EnTeQuAk

**Comment by EnTeQuAk** [comment]
Alrighty, so I looked into this and it's actually quite simple. The file you're looking at - `mozilla.rsa` is a binary file which obviously can be modified but we can't display it.

This is indicated by the `binary` `mime_category` and the mimetype.

![Screenshot from 2019-04-26 10-05-08](https://user-images.githubusercontent.com/139033/56792840-d9875f00-680a-11e9-9a93-fbe75bb5e4b7.png)

So it's perfectly fine to not show any diff here, we also can't fetch actual added/deleted stats properly. In this case, it might be worth showing that extra information as an info message "Binary file, no diff available." or something like that.

**Comment by bobsilverberg** [comment]
Thanks @EnTeQuAk. That seems to be the same reason that I'm seeing a status of `M` for `cose.sig` but no changes. It seems like we do need some extra logic for binary files. I agree that showing "Binary file, no diff available" would be better than showing an empty diff, which is what we do now. I also think we need to possibly incorporate this logic into navigating between diffs, but I'm not totally sure about that.

@kumar303 @willdurand @wagnerand Do you think, when using a keyboard shortcut to navigate between changes, we should just skip any binary files, for which we cannot display a diff, or should they be included along with the a displayed message along the lines of "Binary file, no diff available"? If we do the latter then that will flag to a reviewer that a binary file did in fact change in case that is something they need to know.

**Comment by willdurand** [comment]
> Do you think, when using a keyboard shortcut to navigate between changes, we should just skip any binary files, for which we cannot display a diff

That's a question for @wagnerand.

> should they be included along with the a displayed message along the lines of "Binary file, no diff available"? 

We do need to detect binary files and display a friendly message. Same for images actually. This has not been done yet because of the MVP goal, but maybe it's time to start implementing this detection now (in both browse and compare pages).;

**Comment by bobsilverberg** [comment]
> We do need to detect binary files and display a friendly message. Same for images actually. This has not been done yet because of the MVP goal, but maybe it's time to start implementing this detection now (in both browse and compare pages).

Agreed. I actually started working on displaying images during Q1, but I stopped because we removed it from the MVP. I can pick that up again after I finish off with keyboard shortcuts.

**Comment by kumar303** [comment]
> Do you think, when using a keyboard shortcut to navigate between changes, we should just skip any binary files

The navigation should not skip binary files. We don't want reviewers to miss any changes to binaries (especially for files related to [`connectNative()`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/connectNative)). However, we may want to make some exceptions for known signature files -- this could come in a future optimization. We'll need to be very careful about skipping binary files.

**Comment by wagnerand** [comment]
> > Do you think, when using a keyboard shortcut to navigate between changes, we should just skip any binary files
> 
> The navigation should not skip binary files. We don't want reviewers to miss any changes to binaries (especially for files related to [`connectNative()`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/connectNative)). However, we may want to make some exceptions for known signature files -- this could come in a future optimization. We'll need to be very careful about skipping binary files.

I agree. I would start with not skipping any changed files We can carefully consider making exceptions in follow-ups when we identify them and think they prove enough gain.

**Comment by kumar303** [comment]
Related: it's not possible to view the diff for a deleted file: https://github.com/mozilla/addons-server/issues/11364

**Comment by AlexandraMoga** [comment]
Verifed on -dev with Fx66, Win10x64

Files with an `added` or `modified` status will be highlighted in light green and `deleted` files are pink-ish (similar to the colors used by the file viewer)
At hover, there is also a tooltip that specifies the actual status of the file

@kumar303 
This is an enhancement only, but I think we should make the distinction more clear, especially for `added` and `modified` files which both are highlighted in green. Maybe adding the initial of each status to changed files is a good idea - i.e `A` for `added`, `M` for `modified` etc. I can file a follow up for this.

![2019-05-21_18-38-15](https://user-images.githubusercontent.com/31961530/58111303-e4f75b80-7bf9-11e9-802c-1202a7b82b76.gif)

**Comment by kumar303** [comment]
> Maybe adding the initial of each status to changed files is a good idea

Sure, it's worth filing as a follow-up since this came up in review as well. I'd like to get some feedback from the users first but it's still worth filing.
